kong/runloop/handler.lua

```diff
@@ -435,13 +456,45 @@ return {
         log(ngx.CRIT, "could not ensure API router is up to date: ", err)
 
       elseif api_router_version ~= version then
-        log(DEBUG, "rebuilding API router")
 
-        local ok, err = build_api_router(singletons.dao, version)
+        -- wrap api_router rebuilds in a per-worker mutex (via ngx.semaphore)
+        -- this prevents dogpiling the database during rebuilds in
+        -- high-concurrency traffic patterns
+        -- requests that arrive on this process during a api_router rebuild will be
+        -- queued. once the semaphore resource is acquired we re-check the
+        -- api_router version again to prevent unnecessary subsequent rebuilds
+        --
+        -- note the ten-second wait() timeout. we should be able to safely assume
+        -- that a api_router rebuild event should never take this long
+
+        local ok, err = build_api_router_semaphore:wait(10)
         if not ok then
-          api_router_err = err
-          log(ngx.CRIT, "could not rebuild API router: ", err)
+          return responses.send_HTTP_INTERNAL_SERVER_ERROR(
+            "error attempting to acquire build_api_router lock: " .. err
+          )
         end
+
+        -- lock acquired but we might not need to rebuild the api_router (if we were
+        -- not the first request in this process to enter this code path)
+        -- check again and rebuild if necessary
+
+        version, err = cache:get("api_router:version", CACHE_ROUTER_OPTS, utils.uuid)
+        if err then
+          log(ngx.CRIT, "could not ensure api_router is up to date: ", err)
+
+        elseif api_router_version ~= version then
+
```

Comment:
Mind getting rid of this blank line (in both Routes and APIs code paths)?

Reply:
done!